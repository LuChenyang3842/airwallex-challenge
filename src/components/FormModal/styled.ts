import styled from '@emotion/styled';
import { DarkButton } from 'src/styledComponents/Buttons';

export const SubmitButton = styled(DarkButton)`
  width: 100%;
`;
