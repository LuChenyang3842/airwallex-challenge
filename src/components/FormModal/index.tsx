import { Alert, Form, Modal } from 'antd';
import Input from 'antd/lib/input/Input';
import React, { useState } from 'react';
import { requestInvite } from 'src/service/requestInvite';
import { ModelTitle } from 'src/styledComponents/ModalTitle';
import SuccModal from 'src/components/SuccModal';
import { SubmitButton } from './styled';

interface IFormModalProps {
  visible: boolean;
  handleOk: () => void;
  handleCancel: () => void;
}
interface IFormData {
  fullname: string;
  email: string;
  confirmEmail: string;
}

export const ErrorWording = {
  requireName: 'Full name is required',
  requireEmail: 'Email is required',
  requireConfirm: 'Confirm email is required',
  invalidName: 'Name must be more then 3 characters',
  invalidEmail: 'Please enter a valid email!',
  invalidConfirm: 'The two emails that you entered do not match',
};

const FormModal: React.FC<IFormModalProps> = (props) => {
  const { visible, handleCancel, handleOk } = props;
  const [form] = Form.useForm();
  const [loading, setLoading] = useState(false);
  const [succModalVisible, setSuccModaltVisible] = useState<boolean>(false);
  const [errorMsg, setErrorMsg] = useState('');
  const onFinish = async (values: IFormData) => {
    try {
      await requestInvite({
        name: values.fullname,
        email: values.email,
      });
      handleOk();
      setSuccModaltVisible(true);
    } catch (e: any) {
      setErrorMsg(e?.data.errorMessage ?? '');
    } finally {
      setLoading(false);
    }
  };
  // submit form
  const onClickSubmit = () => {
    setErrorMsg('');
    setLoading(true);
    form.submit();
  };

  const onFinishFailed = () => {
    setLoading(false);
  };

  const handleCloseSuccModal = () => {
    setSuccModaltVisible(false);
  };

  return (
    <>
      <SuccModal
        visible={succModalVisible}
        handleClose={handleCloseSuccModal}
      ></SuccModal>
      <Modal
        title={<ModelTitle> Request an invite</ModelTitle>}
        visible={visible}
        onOk={handleOk}
        onCancel={handleCancel}
        centered={true}
        closable={true}
        destroyOnClose={true}
        footer={null}
      >
        <Form
          form={form}
          layout="horizontal"
          onFinish={onFinish}
          onFinishFailed={onFinishFailed}
        >
          <Form.Item
            name="fullname"
            rules={[
              { required: true, message: ErrorWording.requireName },
              () => ({
                validator(_, value) {
                  if (!value || value.trim().length >= 3) {
                    return Promise.resolve();
                  }
                  return Promise.reject(new Error(ErrorWording.invalidName));
                },
              }),
            ]}
          >
            <Input placeholder="Full Name" data-testid="fullname-input" />
          </Form.Item>
          <Form.Item
            name="email"
            rules={[
              { required: true, message: ErrorWording.requireEmail },
              {
                type: 'email',
                message: ErrorWording.invalidEmail,
              },
            ]}
          >
            <Input placeholder="Email" data-testid="email-input" />
          </Form.Item>
          <Form.Item
            name="confirmEmail"
            rules={[
              { required: true, message: ErrorWording.requireConfirm },
              ({ getFieldValue }) => ({
                validator(_, value) {
                  if (!value || getFieldValue('email') == value) {
                    return Promise.resolve();
                  }
                  return Promise.reject(new Error(ErrorWording.invalidConfirm));
                },
              }),
            ]}
          >
            <Input placeholder="Confirm Email" data-testid="confirm-input" />
          </Form.Item>
          <Form.Item>
            <SubmitButton
              loading={loading}
              type="primary"
              size="large"
              onClick={onClickSubmit}
              data-testid="submit"
            >
              Submit
            </SubmitButton>
          </Form.Item>
          {errorMsg && <Alert type="error" message={errorMsg} showIcon />}
        </Form>
      </Modal>
    </>
  );
};

export default FormModal;
