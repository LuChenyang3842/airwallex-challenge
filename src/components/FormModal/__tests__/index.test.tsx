import { render, fireEvent, cleanup, waitFor } from '@testing-library/react';
import FormModal, { ErrorWording } from '..';
import { screen } from '@testing-library/dom';
import { act } from 'react-dom/test-utils';

describe('From Modal test', () => {
  let visible = true;
  const handleOk = jest.fn();
  const handleCancel = jest.fn();
  const mockEmailA = 'mockEmailA@email.com';
  const mockEmailB = 'mockEmailB@email.com';
  const mockInvalidEmail = 'a';
  const mockValidName = 'mockName';
  const mockInvalidName = 'i';
  const inputByTestId = (input: string, testid: string) => {
    const node = screen.getByTestId(testid);
    fireEvent.change(node, { target: { value: input } });
  };

  const inputName = (input: string) => {
    inputByTestId(input, 'fullname-input');
  };

  const inputEmail = (input: string) => {
    inputByTestId(input, 'email-input');
  };

  const inputConfirm = (input: string) => {
    inputByTestId(input, 'confirm-input');
  };

  const clickSubmitBtn = () => {
    const submitButton = screen.getByTestId('submit');
    fireEvent.click(submitButton);
  };

  const mockValidInputs = () => {
    inputName(mockValidName);
    inputEmail(mockEmailA);
    inputConfirm(mockEmailA);
  };

  beforeAll(() => {
    // reference: https://github.com/ant-design/ant-design/issues/21096
    Object.defineProperty(window, 'matchMedia', {
      value: () => {
        return {
          matches: false,
          // eslint-disable-next-line @typescript-eslint/no-empty-function
          addListener: () => {},
          // eslint-disable-next-line @typescript-eslint/no-empty-function
          removeListener: () => {},
        };
      },
    });
  });
  beforeEach(() => {
    visible = true;
  });

  afterEach(cleanup);

  it('Form Modal renders correctly -- close', () => {
    visible = false;
    const { baseElement } = render(
      <FormModal
        visible={visible}
        handleOk={handleOk}
        handleCancel={handleCancel}
      />,
    );
    expect(baseElement).toMatchSnapshot();
  });

  it('handle cancel correctly invoked', () => {
    const { baseElement } = render(
      <FormModal
        visible={visible}
        handleOk={handleOk}
        handleCancel={handleCancel}
      />,
    );

    const buttonsClose = baseElement.getElementsByClassName('ant-modal-close');
    expect(buttonsClose.length).toBe(1);
    act(() => {
      fireEvent.click(buttonsClose[0]);
    });
    expect(handleCancel).toBeCalledTimes(1);
  });

  it('error message display - empty submission', async () => {
    const { baseElement } = render(
      <FormModal
        visible={visible}
        handleOk={handleOk}
        handleCancel={handleCancel}
      />,
    );
    act(() => {
      clickSubmitBtn();
    });
    await waitFor(() => {
      const erroMsgs = baseElement.getElementsByClassName(
        'ant-form-item-explain-error',
      );
      expect(erroMsgs.length).toBe(3);
      expect(erroMsgs[0].textContent).toBe(ErrorWording.requireName);
      expect(erroMsgs[1].textContent).toBe(ErrorWording.requireEmail);
      expect(erroMsgs[2].textContent).toBe(ErrorWording.requireConfirm);
    });
  });

  it('error message display - invalid name', async () => {
    const { baseElement } = render(
      <FormModal
        visible={visible}
        handleOk={handleOk}
        handleCancel={handleCancel}
      />,
    );
    act(() => {
      mockValidInputs();
      inputName(mockInvalidName);
      clickSubmitBtn();
    });
    await waitFor(() => {
      const erroMsgs = baseElement.getElementsByClassName(
        'ant-form-item-explain-error',
      );
      expect(erroMsgs.length).toBe(1);
      expect(erroMsgs[0].textContent).toBe(ErrorWording.invalidName);
    });
  });

  it('error message display - invalid email', async () => {
    const { baseElement } = render(
      <FormModal
        visible={visible}
        handleOk={handleOk}
        handleCancel={handleCancel}
      />,
    );
    act(() => {
      inputName(mockValidName);
      inputEmail(mockInvalidEmail);
      inputConfirm(mockInvalidName);
      clickSubmitBtn();
    });
    await waitFor(() => {
      const erroMsgs = baseElement.getElementsByClassName(
        'ant-form-item-explain-error',
      );
      expect(erroMsgs.length).toBe(2);
      expect(erroMsgs[0].textContent).toBe(ErrorWording.invalidEmail);
      expect(erroMsgs[0].textContent).toBe(ErrorWording.invalidEmail);
    });
  });

  it('error message display - invalid confirm', async () => {
    const { baseElement } = render(
      <FormModal
        visible={visible}
        handleOk={handleOk}
        handleCancel={handleCancel}
      />,
    );
    inputName(mockValidName);
    inputEmail(mockEmailA);
    inputConfirm(mockEmailB);
    clickSubmitBtn();
    await waitFor(() => {
      const erroMsgs = baseElement.getElementsByClassName(
        'ant-form-item-explain-error',
      );
      expect(erroMsgs.length).toBe(1);
      expect(erroMsgs[0].textContent).toBe(ErrorWording.invalidConfirm);
    });
  });
});
