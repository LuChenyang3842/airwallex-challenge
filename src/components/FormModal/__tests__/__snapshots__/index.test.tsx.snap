// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`From Modal test Form Modal renders correctly -- close 1`] = `
<body>
  <div />
</body>
`;
